File: python/src/adbc_example.py
"""
This is an example for how to query the Semantic Layer using the ADBC driver.

This example is meant to be run from the CLI when you have the env var DBT_JDBC_URL
set. See the README for details.

For using in a Jupyter notebook or your own program, you should only need the `execute_query`
function, and you need to adapt it to your needs. In this example we simply print out the
DataFrame, but you probably want to return it instead.
"""
import os
import sys
from dataclasses import dataclass
from urllib.parse import parse_qs, urlparse

from adbc_driver_flightsql import DatabaseOptions
from adbc_driver_flightsql.dbapi import connect


@dataclass
class ConnAttr:
    host: str  # "grpc+tls:semantic-layer.cloud.getdbt.com:443"
    environment_id: str  # 42
    token: str  # dbts_thisismyprivateservicetoken


def parse_jdbc_uri(uri):
    """Helper function to convert the JDBC url into ConnAttr."""
    parsed = urlparse(uri)
    params = {k.lower(): v[0] for k, v in parse_qs(parsed.query).items()}
    return ConnAttr(
        host=parsed.path.replace("arrow-flight-sql", "grpc")
        if params.pop("useencryption", None) == "false"
        else parsed.path.replace("arrow-flight-sql", "grpc+tls"),
        environment_id=params.pop('environmentid'),
        token=params.pop('token'),
    )


def execute_query(host, environment_id, token, query, db_opts=None):
    """Execute a Semantic Layer query.

    host must be in a fromat like: grpc+tls:semantic-layer.cloud.getdbt.com:443
    db_opts is a dictionary of additional DB options to pass in
    """
    opts = db_opts or {}

    with connect(
        conn_attr.host,
        db_kwargs={
            DatabaseOptions.AUTHORIZATION_HEADER.value: f"Bearer {token}",
            f"{DatabaseOptions.RPC_CALL_HEADER_PREFIX.value}environmentid": environment_id,
            DatabaseOptions.WITH_COOKIE_MIDDLEWARE.value: "true",
            **opts
        },
    ) as conn, conn.cursor() as cur:
        cur.execute(query)
        df = cur.fetch_df()  # fetches as Pandas DF, can also do fetch_arrow_table
    print(df.to_string())


if __name__ == "__main__":
    query = sys.argv[1]
    conn_attr = parse_jdbc_uri(os.environ["DBT_JDBC_URL"])

    execute_query(conn_attr.host, conn_attr.environment_id, conn_attr.token, query)

File: r/adbc_example.R
# This is an example for how to query the Semantic Layer using the ADBC driver.
#
# This example is meant to be run from the CLI when you have the env var DBT_JDBC_URL
# set. See the README for details.
#
# You will need to install the arrow package and the adbcflightsql package by uncommenting
# the lines below before this example will work.
#
# You will likely need to adapt this for your own usage.

# install.packages("arrow")
# install.packages("pak")
# pak::pak("apache/arrow-adbc/r/adbcflightsql")

library(adbcdrivermanager)
library(adbcflightsql)
library(arrow)
library(httr2)

query <- commandArgs(trailingOnly=TRUE)[1] # "select * from {{ semantic_layer.query(metrics=['order_total'], group_by=['metric_time']) }}"

parsed_jdbc_url <- url_parse(Sys.getenv("DBT_JDBC_URL"))

host <- gsub("arrow-flight-sql", "grpc+tls", parsed_jdbc_url$path) # "grpc+tls:semantic-layer.cloud.getdbt.com:443"
environmentId <- parsed_jdbc_url$query$environmentId # "42"
token <- parsed_jdbc_url$query$token # "dbts_thisismyprivateservicetoken"

db <- adbcdrivermanager::adbc_database_init(
  adbcflightsql::adbcflightsql(),
  uri = host,
  adbc.flight.sql.authorization_header = paste("Bearer ", token),
  adbc.flight.sql.rpc.call_header.environmentid = environmentId,
  adbc.flight.sql.rpc.with_cookie_middleware = "true"
)
con <- adbcdrivermanager::adbc_connection_init(db)

df <- con |>
  read_adbc(query) |>
  tibble::as_tibble()

print(df)

adbc_connection_release(con)
adbc_database_release(db)
